components/script/dom/servoparser/html.rs

```diff
@@ -136,58 +141,57 @@ impl<'a> Serialize for &'a Node {
+                    if let Some(head) = stack.last_mut() {
+                        head.1 -= 1;
+                    }
+                },
+                NodeTypeId::DocumentFragment => {},
+                NodeTypeId::Document(_) => panic!("Can't serialize Document node itself"),
+            };
+            while {
+                let mut n_children = 1;
+                if let Some(head) = stack.last() {
+                    n_children = head.1;
                 }
-
-                let children = if let Some(tpl) = node.downcast::<HTMLTemplateElement>() {
-                    // https://github.com/w3c/DOM-Parsing/issues/1
-                    tpl.Content().upcast::<Node>().children()
-                } else {
-                    node.children()
-                };
-
-                for handle in children {
-                    try!((&*handle).serialize(serializer, IncludeNode));
-                }
-
-                if traversal_scope == IncludeNode {
-                    try!(serializer.end_elem(name.clone()));
-                }
-                Ok(())
-            },
-
-            (ChildrenOnly, NodeTypeId::Document(_)) => {
-                for handle in node.children() {
-                    try!((&*handle).serialize(serializer, IncludeNode));
+                n_children == 0
+            } {
+                try!(serializer.end_elem(stack.pop().unwrap().0));
+                if let Some(mut head) = stack.last_mut() {
+                    head.1 -= 1;
                 }
-                Ok(())
-            },
-
-            (ChildrenOnly, _) => Ok(()),
-
-            (IncludeNode, NodeTypeId::DocumentType) => {
-                let doctype = node.downcast::<DocumentType>().unwrap();
-                serializer.write_doctype(&doctype.name())
-            },
-
-            (IncludeNode, NodeTypeId::CharacterData(CharacterDataTypeId::Text)) => {
-                let cdata = node.downcast::<CharacterData>().unwrap();
-                serializer.write_text(&cdata.data())
-            },
-
-            (IncludeNode, NodeTypeId::CharacterData(CharacterDataTypeId::Comment)) => {
-                let cdata = node.downcast::<CharacterData>().unwrap();
-                serializer.write_comment(&cdata.data())
-            },
-
-            (IncludeNode, NodeTypeId::CharacterData(CharacterDataTypeId::ProcessingInstruction)) => {
-                let pi = node.downcast::<ProcessingInstruction>().unwrap();
-                let data = pi.upcast::<CharacterData>().data();
-                serializer.write_processing_instruction(&pi.target(), &data)
-            },
-
-            (IncludeNode, NodeTypeId::DocumentFragment) => Ok(()),
-
-            (IncludeNode, NodeTypeId::Document(_)) => panic!("Can't serialize Document node itself"),
+            }
+        }
+        if stack.is_empty() {
+            Ok(())
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, "Stack is not empty!!!!"))
```

Comment:
What's that for? If this isn't empty, this is a logic error and should probably panic instead.